man/man1/wload.1
.TH WLOAD 1 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wload \- system load average display for W
.SH SYNOPSIS
.nf
wload [-h] [-g \fIgeometry\fP] [-u \fIseconds\fP]
.fi
.SH DESCRIPTION
The
.I wload
program displays a periodically updating
histogram of the system load average.
It is based on
.I xload
for the X window system.
.SH OPTIONS
.I wload
accepts the following options:
.TP
.B \-h
Shows usage line
.TP
.B \-g \fIgeometry\fP
This option is only partially implemented. It currently only supports
the position part of the option to place the window. Implementing the
other parts would require major changes.
.TP
.B \-u \fIseconds\fP
This option specifies the interval in seconds at which \fIwload\fP
updates its display. The minimum amount of time allowed between updates
is 1 second. The default is 10.
.SH AUTHOR
.I wload
was written by Benjamin Lorenz for use with
.IR W (1),
the MiNT window manager. Thanks to Torsten Scherer
(itschere@techfak.uni-bielefeld.de)!
.SH "SEE ALSO"
.IR wserver (1),
.IR wclock (1),
.IR wbiff (1),
.IR wterm (1)
.SH BUGS
should be reported to benni@phil15.uni-sb.de, thanks...
.SH LAST CHANGED
12/95 by TeSche, itschere@techfak.uni-bielefeld.de


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


apps/wload.c
/*
 * wload.c, a part of the W Window System
 *
 * Copyright (C) 1994 by Benjamin Lorenz and
 * Copyright (C) 1994-1997 by Torsten Scherer
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- a system 'load' meter
 *
 * CHANGES:
 * benni 9/94:
 * - Version 1.0
 * ++tesche 95-96:
 * - small bugfix: scale lines were drawn
 *   one pixel too long, fixing this involved more changes in
 *   when to bitblk and so on... :(
 * - okok, another small bugfix, as suggested by Benni
 * - changed for W1R0
 * - added support for Linux68k
 * - major rewrite (-95)
 * - YAABF (-96) :)
 * ++Phx 6/96:
 * - added support for NetBSD
 * ++eero 97-98:
 * - hostname can be > 16 chars, fixed
 * - long options and --parent option
 * - simplified a bit (lines are clipped by server anyway now :))
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#ifdef __NetBSD__
#include <sys/types.h>
#include <sys/param.h>
#include <amiga/vmparam.h>
#include <sys/sysctl.h>
#include <sys/resource.h>
#include <vm/vm_param.h>
#endif
#include <Wlib.h>

#define LFRACTION 2048


static struct option long_options [] = {
  { "geometry", required_argument, NULL, 'g' },
  { "help", no_argument, NULL, 'h' },
  { "parent", required_argument, NULL, 'p'},
  { "seconds", required_argument, NULL, 'u'},
  { NULL, 0, NULL, 0}
};


#ifdef __MINT__

#include <macros.h>
#include <support.h>
#include <mintbind.h>

#elif linux


#define max(a,b) (((a) > (b)) ? (a) : (b))

static void Suptime(ulong *uptime, ulong *load)
{
  static FILE *fp = NULL;
  float f1, f2, f3;

  /* sorry, can't make this static 'cause Linux can't rewind this file
   */
  if (!(fp = fopen("/proc/loadavg", "r"))) {
    fprintf(stderr, "error: can't open /proc/loadavg\n");
    exit(-1);
  }

  fscanf(fp, "%f %f %f\n", &f1, &f2, &f3);
  fclose (fp);

  *uptime = 0; /* we don't need that */
  load[0] = f1 * LFRACTION;
  load[1] = f2 * LFRACTION;
  load[2] = f3 * LFRACTION;
}


#elif __NetBSD__

void Suptime(ulong *uptime, ulong *load)
{
  struct loadavg lavg;
  int mib[2];
  size_t len = sizeof(struct loadavg);
  ulong fscale;
  
  mib[0] = CTL_VM;
  mib[1] = VM_LOADAVG;
  sysctl(mib, 2, &lavg, &len, NULL, 0);  /* fill loadavg structure */
  fscale = (ulong)lavg.fscale;
  *uptime = 0;  /* as for linux ... we don't need that neither? */
  load[0] = ((ulong)lavg.ldavg[0] * LFRACTION) / fscale;
  load[1] = ((ulong)lavg.ldavg[1] * LFRACTION) / fscale;
  load[2] = ((ulong)lavg.ldavg[2] * LFRACTION) / fscale;
}


#else

void Suptime(ulong *uptime, ulong *load)
{
  *uptime = 1;
  load[0] = 1;
  load[1] = 1;
  load[2] = 1;
}

#endif


/*
 * global variables
 */

#define WIN_W	80
#define WIN_H	128

static short winw;
static short winh;
static short header;

static WWIN *win;
static int *loads;


/*
 *
 */

static void w_clear()
{
  w_setmode(win, M_CLEAR);
  w_pbox(win, 0, header, winw, winh - header);
}


static void print_load(int l)
{
  static int x = 0, lceil_old = 1;
  int i, lceil;

  /* insert new load value */
  if (x == winw) {
    for (i=0; i<winw-1; i++) {
      loads[i] = loads[i+1];
    }
    loads[winw - 1] = l;
  } else {
    loads[x] = l;
  }

  /* calc new ceil() value of loads */
  lceil = 0;
  for (i=0; i<winw; i++) {
    if (loads[i] > lceil) {
      lceil = loads[i];
    }
  }
  if ((lceil = (lceil + LFRACTION - 1) / LFRACTION) == 0) {
    lceil = 1;
  }

  if (lceil != lceil_old) {

    /* completely redraw window */
    w_clear();
    w_setmode(win, M_DRAW);
    for (i=0; i<winw; i++) {
      w_vline(win, i, winh - 1, winh - (loads[i] * (winh - header)) / (lceil * LFRACTION));
    }
    w_setmode(win, M_INVERS);
    for (i=1; i<lceil; i++) {
      w_hline(win, 0, winh - (i * (winh - header)) / lceil, winw-1);
    }

    if (x == winw) {
      x--;
    }

  } else {

    if (x >= winw) {
      w_bitblk(win, 1, header, winw-1, winh-header, 0, header);
      x--;
    }

    w_setmode(win, M_CLEAR);
    w_vline(win, x, header, winh-1);
    w_setmode(win, M_DRAW);
    w_vline(win, x, winh - 1, winh  - (loads[x] * (winh - header)) / (lceil * LFRACTION));
    w_setmode(win, M_INVERS);
    for (i=1; i<lceil; i++) {
      w_plot(win, x, winh - (i * (winh - header)) / lceil);
    }
  }

  lceil_old = lceil;
  x++;
}


static void usage (char *name) {
  fprintf (stderr, "\
Usage: %s [OPTION]...\n\
Shows the machine load.\n\
\n\
  -g, --geometry             set size and position\n\
  -u, --seconds              seconds between updates\n\
  -h, --help                 display this help and exit\n\
\n\
", name);
}


int main(int argc, char **argv)
{
  long u = 10000;
  int i, l, c, show_usage = 0;
  ulong parentid = 0;
  ulong uptime;
  ulong load[3];
  ulong avg[3];
  char hostname[32];
  WFONT *font;
  WEVENT *ev;
  short xp = UNDEF;
  short yp = UNDEF;
  WSERVER *wserver;

  while ((c = getopt_long (argc, argv, "g:u:p:h", long_options, NULL)) != -1) {
    switch (c) {

      case 'g':
        scan_geometry (optarg, &winw, &winh, &xp, &yp);
        break;

      case 'u':
	if ((u = atoi(optarg)) < 1) {
	  u = 1;
	}
	if (u > 60) {
	  u = 60;
	}
	u *= 1000;
	break;

      case 'p':
        sscanf (optarg, "%lx", &parentid);
        break;

      default:
        show_usage = 1;
        break;
    }
  }

  if (show_usage) {
    usage (argv[0]);
    exit (0);
  }

  if (!(wserver = w_init())) {
    fprintf(stderr, "error: can't connect to wserver\n");
    exit(1);
  }
  limit2screen(win, &xp, &yp);

  if (parentid) {
    WWIN *parent;
    parent = w_winFromID(parentid);
    winw = parent->width;
    winh = parent->height;
    win = w_createChild (parent, winw, winh, W_NOBORDER);
    xp = yp = 0;
  } else {
    if (winw <= 0) {
      winw = WIN_W;
    }
    if (winh <= 0) {
      winh = WIN_H;
    }
    win = w_create(winw, winh, W_MOVE);
  }
  if (!win) {
    fprintf(stderr, "error: can't create window\n");
    exit(1);
  }

  if (w_open(win, xp, yp)) {
    fprintf(stderr, "error: can't open window\n");
    w_exit();
    exit(1);
  }

  gethostname(hostname, sizeof(hostname)-1);

  if (!(loads = malloc(sizeof(*loads) * winw))) {
    fprintf(stderr, "error: can't open window\n");
    w_exit();
    exit(1);
  }
  for (i = 0; i < winw; i++)
    loads[i] = 0;
  for (i = 0; i < 3; i++)
    avg[i]=0;

  w_setmode(win, M_DRAW);
  if (!(font = w_loadfont("cour", 0, F_BOLD))) {
    font = w_loadfont(NULL, 0, 0);
  }
  w_setfont(win, font);
  w_printstring(win, 1, 1, hostname);
  w_hline(win, 0, winh - 1, winw - 1);
  header = font->height;

  while (1) {
    Suptime(&uptime, load);
    avg[0] = avg[1];
    avg[1] = avg[2];
    avg[2] = load[0];

    /* calc load in 1/LFRACTION fractions  */
    l = (avg[0] + avg[1] + avg[2]) / 3;
    print_load(l);

    /* watch out for `exit' event or 'q' key */
    if ((ev = w_queryevent(NULL, NULL, NULL, u))) {
      if ((ev->type == EVENT_KEY && ev->key == 'q') ||
          (ev->type == EVENT_GADGET && ev->key == GADGET_EXIT)) {
	w_exit();
	exit(0);
      }
    }
  }
}
